feat(assessments): add assessments module and related functionality

tomer-shvadron opened the pull request:
- Introduce AssessmentsControllerExternal to handle assessment endpoints
- Implement AssessmentsService for fetching and creating assessments
- Create DTOs for assessment operations including validation rules


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Introduced KYB & Ownership and Identity Verification assessments with full CRUD and status management via API endpoints.
  - Added React components and hooks for listing, creating, viewing, and updating assessments with pagination, filtering, and search.
  - Implemented dialogs for creating assessments and identity verification checks with form validation and success states.
  - Added status badges, update dialogs, and notes integration for assessment and verification records.
  - Included section observer navigation for detailed assessment pages.
  - Integrated localized toast notifications for creation and status update actions.

- **Refactor**
  - Replaced inline section observer logic with a reusable shared SectionObserver component for better maintainability.

- **Chores**
  - Added new status constants, types, and mappings for checks and assessments.
  - Extended customer configuration flags and routing to support new assessment and verification features.
  - Updated date-time validation in business schemas to use ISO string format.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

chesterkmr committed:
feat: added kyb & ubos boilerplate

chesterkmr committed:
feat: added identity verficaition page boilerplate

chesterkmr committed:
feat: implemented create kyb & ubo check form

chesterkmr committed:
feat: added mocks for kyb & ubos page

chesterkmr committed:
feat: implemented mocked identity verification & fixes

chesterkmr committed:
feat: updated kyb & stakeholders design

chesterkmr committed:
feat: added initial checks module

chesterkmr committed:
feat: fixed checks endpoint

chesterkmr committed:
feat: connected backoffice to kyb & ubo endpoint

chesterkmr committed:
feat: added check creation requests & updated toast

r4zendev committed:
feat: kyb check page

chesterkmr committed:
feat: added feature flags

chesterkmr committed:
feat: updated schemas & added swagger to checks controller

chesterkmr committed:
fix: resolved imports & fixed structure

chesterkmr committed:
fix: finalized table data fetching and mapping

chesterkmr committed:
fix: fixed dtos

chesterkmr committed:
feat: added rendering of merchant id

chesterkmr committed:
chore: disabled findings

chesterkmr committed:
fix: replaced internal with external controller

chesterkmr committed:
fix: reverted main

chesterkmr committed:
fix: fixed fetchers urls

chesterkmr committed:
fix: fixed format

chesterkmr committed:
feat: state is now required only for US

chesterkmr committed:
fix: fixed kyb checks schema

r4zendev committed:
use blocks

chesterkmr committed:
fix: schema fix

r4zendev committed:
upd

chesterkmr committed:
fix: fixed table rendering

tomer-shvadron committed:
refactor(kyb-ubo): improve column definitions and badge rendering

- Simplify column accessor definitions in KybAndUboChecksTable
- Adjust badge rendering to safely access status data properties
- Remove unnecessary console log in KycAndUboLogic hook

r4zendev committed:
upd schema, add data not available placeholders

r4zendev committed:
upd references to check

r4zendev committed:
remove mock

chesterkmr committed:
fix: fixed failed status crash & added polling

chesterkmr committed:
fix: added state to checks payload

r4zendev committed:
remove garbage

r4zendev committed:
fixes

r4zendev committed:
upd

r4zendev committed:
hackathon text fixes

r4zendev committed:
hackathon fixes 2

tomer-shvadron committed:
feat(assessments): add assessments module and related functionality

- Introduce AssessmentsControllerExternal to handle assessment endpoints
- Implement AssessmentsService for fetching and creating assessments
- Create DTOs for assessment operations including validation rules

tomer-shvadron committed:
refactor(app.module): reorganize and standardize module imports

- Rearrange imports for better readability and maintainability
- Ensure consistent naming format for endpoint summaries in Swagger documentation
- Update subproject commit reference in data migrations

tomer-shvadron committed:
refactor(assessments): streamline assessments controller and service

- Remove unused getAssessments method and corresponding DTO
- Introduce methods for retrieving specific and grouped KYB assessments
- Enhance create assessment functionality with business ID tracking

tomer-shvadron committed:
Merge branch 'dev' into bal-4146-incorporate-hackathon-workflows-service-changes-into-dev

tomer-shvadron committed:
feat(assessments): enhance assessments route and schema

- Refactor assessments response to include input and type in results
- Update schema to accommodate pagination parameters and input field
- Improve service logic for jurisdiction code extraction

tomer-shvadron committed:
feat(assessments): add fetching functionality for Kyb and Ownership assessment

- Implement fetchKybAndOwnershipAssessment function to retrieve assessment by ID
- Introduce useKybAndOwnershipAssessmentQuery hook for query handling
- Update relevant components to reflect changes in naming conventions and loading states

tomer-shvadron committed:
Merge branch 'dev' into bal-4146-incorporate-hackathon-workflows-service-changes-into-dev

tomer-shvadron committed:
feat(assessments): update KYB assessment schemas and labels

- Add CreateKybAndOwnershipAssessmentResponseSchema for validation
- Rename 'KYB & UBOs' to 'KYB & Ownership' in sidebar
- Update dialog message to reflect new assessment terminology

tomer-shvadron committed:
refactor(assessments): rename merchant report statuses to assessment statuses

- Replace occurrences of MERCHANT_REPORT_STATUSES with ASSESSMENT_STATUSES
- Update related components and imports for consistency
- Remove deprecated code related to merchant report statuses

tomer-shvadron committed:
feat(status): add assessment status update functionality

- Introduce new toast messages for assessment status updates
- Implement updateAssessmentStatus in the assessment service
- Refactor component to handle assessment ID instead of report ID

tomer-shvadron committed:
chore(release): bump versions and update dependencies

- Bump versions across multiple packages to reflect latest changes
- Update dependencies for @ballerine/common, workflow-browser-sdk, and workflow-node-sdk

r4zendev committed:
add checks section

tomer-shvadron committed:
feat(assessments): enhance Kyb and Ownership assessments functionality

- Import ASSESSMENT_STATUSES_MAP for status handling
- Update pagination to use nested page object structure
- Introduce validation schema for assessments query parameters

tomer-shvadron committed:
refactor(kybAndOwnership): remove unused customer query

- Eliminate the use of the customer query hook as it is no longer needed
- Clean up code by removing related imports and state resets

tomer-shvadron committed:
Merge branch 'dev' into bal-4146-incorporate-hackathon-workflows-service-changes-into-dev

Omri-Levy commented on apps/backoffice-v2/public/locales/en/toast.json:
```diff
@@ -151,5 +151,14 @@
     "error": "Error occurred while transitioning to edit collection flow."
+  },
+  "kyb_and_ownership_assessment_creation": {
+    "success": "KYB & Ownership assessment created successfully.",
```
Not sure we want to mix internal terminology with product. Assessment might not be clear enough for a customer.

tomer-shvadron replied:
I put down a comment to myself for the sign-off with Adi and Nitzan

tomer-shvadron committed:
lock file fix

tomer-shvadron committed:
lock file fix

Omri-Levy commented on apps/backoffice-v2/src/common/components/organisms/SectionObserver/SectionObserver.tsx:
```diff
@@ -0,0 +1,216 @@
+  onSidebarToggle?: (isOpen: boolean) => void;
+};
+
+export const SectionObserver = <TSection extends Section>({
```
Considering we did not add tests I'd suggest we use a package for this whether it is a hook or a component.

Omri-Levy commented on apps/backoffice-v2/src/domains/assessments/fetchers.ts:
```diff
@@ -0,0 +1,209 @@
+import qs from 'qs';
+import { ASSESSMENT_STATUSES, ASSESSMENT_STATUSES_MAP } from '@ballerine/common';
+
+export const KybAndOwnershipAssessmentStatusSchema = z.enum([
+  'pending',
+  'approved',
+  'rejected',
+  'in-progress',
+]);
```
This enum is already in common right?

Omri-Levy commented on apps/backoffice-v2/src/domains/assessments/fetchers.ts:
```diff
@@ -0,0 +1,209 @@
+    timeout: 30_000,
+  });
+
+  if (error) {
+    return handleZodError(error, result);
+  }
```
Is this if added considering the logic inside handleZodError ?

Omri-Levy commented on apps/backoffice-v2/src/domains/assessments/hooks/mutations/useCreateKybAndOwnershipAssessmentMutation/useCreateKybAndOwnershipAssessmentMutation.ts:
```diff
@@ -0,0 +1,44 @@
+
+      toast.error(
+        t(`toast:kyb_and_ownership_assessment_creation.error`, {
+          errorMessage: isObject(error) && 'message' in error ? error.message : error,
```
We also have a util called isErrorWithMessage.

Omri-Levy commented on apps/backoffice-v2/src/domains/customer/fetchers.ts:
```diff
@@ -43,6 +43,8 @@ const CustomerSchema = z.object({
       isDemoAccount: z.boolean().default(false),
       isDemoKybEnabled: z.boolean().default(false),
+      createKybAndOwnershipAssessment: z.boolean().default(true),
+      createIdentityVerification: z.boolean().default(true),
```
We both break from our naming conventions and it looks as true creates something as opposed to a feature or some UI being enabled.

tomer-shvadron committed:
chore(release): bump versions and update dependencies

- Bump versions for multiple packages to include new updates
- Update dependencies across various components to ensure compatibility

Omri-Levy commented on apps/backoffice-v2/src/pages/KybAndOwnership/components/CreateKybAndOwnershipAssessmentDialog/CreateKybAndOwnershipAssessmentDialog.tsx:
```diff
@@ -0,0 +1,255 @@
+
+export const CreateKybAndOwnershipAssessmentDialog = ({
+  disabled,
+  children,
```
Children going into the trigger is unintuitive. I think having a prop for trigger or using children for the dialog content would be more common.

Omri-Levy commented on apps/backoffice-v2/src/pages/KybAndOwnership/components/CreateKybAndOwnershipAssessmentDialog/CreateKybAndOwnershipAssessmentDialog.tsx:
```diff
@@ -0,0 +1,255 @@
+                  </FormItem>
+                )}
+              />
+              <FormField
+                control={form.control}
+                name="country"
```
There is an opportunity here to use the latest country component and create one for state.

Omri-Levy commented on apps/backoffice-v2/src/pages/KybAndOwnership/components/KybAndOwnershipAssessmentStatus/KybAndOwnershipAssessmentStatus.tsx:
```diff
@@ -0,0 +1,201 @@
+
+    mutateUpdateAssessmentStatus({ assessmentId, status: dialogState.status });
+
+    closeDialog();
```
This probably should be on the mutation's onSuccess. Both the form.reset and the closeDialog.

Omri-Levy commented on apps/backoffice-v2/src/pages/KybAndOwnership/components/KybAndOwnershipAssessmentStatus/KybAndOwnershipAssessmentStatus.tsx:
```diff
@@ -0,0 +1,201 @@
+    [isUpdatingAssessmentStatus, status],
+  );
+
+  if (!status || !assessmentId) {
+    return null;
+  }
```
There are many instances in this PR where a component is hid instead of having a fallback state and a skeleton. Worth acknowledging how it would affect UX and layout shift.

Omri-Levy commented on apps/backoffice-v2/src/pages/KybAndOwnership/components/KybAndOwnershipAssessmentStatus/KybAndOwnershipAssessmentStatusBadge.tsx:
```diff
@@ -0,0 +1,80 @@
+}) => {
+  const isInProgress = status === ASSESSMENT_STATUSES_MAP['in-progress'];
+
+  const { ref, styles } = useEllipsesWithTitle<HTMLSpanElement>();
```
Newer areas in our system use the tooltip component from ShadCN.

Omri-Levy commented on apps/backoffice-v2/src/pages/KybAndOwnership/components/KybAndOwnershipAssessmentStatus/KybAndOwnershipAssessmentStatusButton.tsx:
Feels like this component would be better off with composition than being a component of its own. I'd argue there's no complexity to abstract here or much potential for re-use.

Omri-Levy commented on apps/backoffice-v2/src/pages/KybAndOwnership/components/KybAndOwnershipAssessmentStatus/fetchers.ts:
I'd expect this fetcher to be in the assessments domain and not co-located to the status UI component.

Omri-Levy commented on apps/backoffice-v2/src/pages/KybAndOwnership/components/KybAndOwnershipAssessmentStatus/helpers/get-base-note-content.ts:
Worth noting we have this identical component repeated multiple times even though it's generic.

Omri-Levy commented on apps/backoffice-v2/src/pages/KybAndOwnership/components/KybAndOwnershipAssessmentStatus/helpers/get-note-content-for-unsubscribe.ts:
Duplication like mentioned in get-base-note-content comment.

Omri-Levy commented on apps/backoffice-v2/src/pages/KybAndOwnership/components/KybAndOwnershipAssessmentStatus/hooks/useUpdateKybAndOwnershipAssessmentStatus/useUpdateKybAndOwnershipAssessmentStatus.tsx:
```diff
@@ -0,0 +1,49 @@
+import { HttpError } from '@/common/errors/http-error';
+import { updateAssessmentStatus } from '../../fetchers';
+
+export const useUpdateKybAndOwnershipAssessmentStatus = ({
```
Same as the fetchers, I think having this mutation under the assessments domain makes sense.
```suggestion
export const useUpdateKybAndOwnershipAssessmentStatusMutation = ({
```

Omri-Levy commented on apps/backoffice-v2/src/pages/KybAndOwnership/components/KybAndOwnershipAssessmentsTable/columns.tsx:
```diff
@@ -0,0 +1,115 @@
+      }),
+    ];
+
+    return columns.filter(column => {
+      const meta = column.meta;
+
+      if (meta?.conditional) {
+        return meta.showColumn;
+      }
+
+      return true;
+    });
```
Tanstack table has a first-class API for conditional columns. https://tanstack.com/table/latest/docs/guide/column-visibility#column-visibility-state

Omri-Levy commented on apps/backoffice-v2/src/pages/KybAndOwnership/hooks/useKybAndOwnershipLogic/useKybAndOwnershipLogic.tsx:
```diff
@@ -0,0 +1,138 @@
+  );
+
+  const open = isCreating ?? false;
+  const toggleOpen = (value?: boolean) => setSearchParams({ isCreating: value });
+
+  const reportQuery = {
+    page: {
+      number: page,
+      size: pageSize,
+    },
+    from,
+    to,
+  };
```
Consider useCallback and useMemo

Omri-Levy commented on apps/backoffice-v2/src/pages/KybAndOwnershipAssessmentPage/KybAndOwnershipAssessmentPage.tsx:
```diff
@@ -0,0 +1,130 @@
+
+  const sectionRefs = useRef<{ [key: string]: HTMLDivElement | null }>({});
+
+  if (isLoadingAssessment || !assessment) {
```
If there is no assessment we'll end up with an infinite skeleton.

Omri-Levy commented on apps/backoffice-v2/src/pages/KybAndOwnershipAssessmentPage/hooks/useKybAndOwnershipAssessmentPageLogic.tsx:
```diff
@@ -0,0 +1,367 @@
+  ): CheckItem => {
+    return {
+      displayName,
+      status: condition === true ? 'positive' : condition === false ? 'negative' : 'neutral',
+      note: condition === true ? positiveNote : condition === false ? negativeNote : neutralNote,
```
Hard to read and understand. Consider simplifying to `condition ? 'positive' : 'negative'` and handling non-boolean separately or even better not using ternary here could prove simpler.

Omri-Levy commented on apps/backoffice-v2/src/pages/KybAndOwnershipAssessmentPage/hooks/useKybAndOwnershipAssessmentPageLogic.tsx:
```diff
@@ -0,0 +1,367 @@
+
+  // 1. Registry Information check
+  checks.push(
+    createCheck(
+      'Registry Information',
+      assessment?.companyRegistryInformation
+        ? assessment.companyRegistryInformation.status === 'completed' && Boolean(registryData)
+        : null,
+      'Extracted',
+      'Flagged',
+    ),
```
Any reason not to inline the createCheck calls into the array at its declaration?

Omri-Levy commented on apps/backoffice-v2/src/pages/KybAndOwnershipAssessmentPage/hooks/useKybAndOwnershipAssessmentPageLogic.tsx:
```diff
@@ -0,0 +1,367 @@
+
+  const onNavigateBack = useCallback(() => navigate(-1), [navigate]);
+
+  const notes = [] as any[];
+  // TODO: Implement notes when API is available
+  // const { data: notes } = useNotesByNoteable({
+  //   noteableId: assessment?.id || '',
+  //   noteableType: 'Report',
+  // });
```
Is this supposed to be deleted?

Omri-Levy commented on apps/backoffice-v2/src/pages/KybAndOwnershipAssessmentPage/hooks/useKybAndOwnershipAssessmentPageLogic.tsx:
```diff
@@ -0,0 +1,367 @@
+                    key={`${check.displayName}-${index}`}
+                    className={ctw(
+                      'flex h-16 items-center justify-between rounded-md border border-gray-200 px-4',
+                      check.status === 'positive' && 'bg-green-50',
+                      check.status === 'negative' && 'bg-red-50',
```
Is this supposed to be destructive and success for colors?

Omri-Levy commented on apps/backoffice-v2/src/pages/KybAndOwnershipAssessmentPage/hooks/useKybAndOwnershipAssessmentPageLogic.tsx:
```diff
@@ -0,0 +1,367 @@
+            ) : (
+              <Card>
+                <CardContent className="p-6">
+                  {assessment?.companyRegistryInformation?.status === 'failed'
+                    ? 'Company Registry Information check failed'
+                    : 'Company Registry Information data is not available'}
```
We might want to write somewhere that as part of my checks errors task these errors are on the errors column of the check record.

Omri-Levy commented on apps/backoffice-v2/src/router.tsx:
```diff
@@ -101,6 +107,46 @@ const router = createBrowserRouter([
               },
+              {
+                loader: async () => {
+                  await queryClient.ensureQueryData(customerQueryKeys.getCurrent());
```
We'll probably need to disable this query if the user is not authenticated.

Omri-Levy commented on services/workflows-service/src/assessments/assessments.service.ts:
```diff
@@ -0,0 +1,84 @@
+      );
+
+      return result.data;
+    } catch (error) {
+      throw new InternalServerErrorException(error);
+    }
```
Let's discuss this face-to-face.

Omri-Levy commented on services/workflows-service/src/assessments/dtos/create-assessment.dto.ts:
```diff
@@ -0,0 +1,50 @@
+    description: 'Type of assessment',
+    example: 'kyb_and_ownership',
+  })
+  @IsString()
+  @oneOf(['kyb_and_ownership'])
```
oneOf is probably the only one needed and not IsString.

Omri-Levy commented on services/workflows-service/src/assessments/dtos/create-assessment.dto.ts:
```diff
@@ -0,0 +1,50 @@
+    description: 'Business ID',
+    example: '1234567890',
+  })
+  @IsString()
+  @IsOptional()
```
Validating for non-empty string is also worthwhile.

Omri-Levy commented on services/workflows-service/src/assessments/dtos/get-kyb-and-ownership-assessments.dto.ts:
```diff
@@ -0,0 +1,45 @@
+
+export const GetKybAndOwnershipAssessmentsSchema = z.object({
+  page: z.object({
+    number: z.coerce.number().int().positive(),
+    size: z.coerce.number().int().positive().max(100),
```
Great chain. 💪🏼

Omri-Levy commented on services/workflows-service/src/common/dto.ts:
```diff
@@ -24,11 +24,11 @@ export const validateOrderBy = (value: unknown, validColumns: readonly string[])
-  @ApiProperty({ name: 'page[number]', example: 1 })
+  @ApiProperty({ name: 'page[number]', example: 1, default: 1 })
   @Transform(({ value }) => parseInt(value, 10))
   number!: number;
```
```suggestion
  number!: number = 1;
```

Omri-Levy commented on services/workflows-service/src/common/dto.ts:
```diff
@@ -24,11 +24,11 @@ export const validateOrderBy = (value: unknown, validColumns: readonly string[])
-  @ApiProperty({ name: 'page[size]', example: 20 })
+  @ApiProperty({ name: 'page[size]', example: 20, default: 20 })
   @Transform(({ value }) => parseInt(value, 10))
   size!: number;
```
```suggestion
  size!: number = 20;
```
